src/extension.ts

```diff
@@ -41,110 +40,152 @@ export function activate(context: vscode.ExtensionContext) {
-						if (selected === "Reload") {
-							vscode.commands.executeCommand("workbench.action.reloadWindow");
-						}
-					});
-				} else {
-					return;
+					const selected = await vscode.window.showInformationMessage(reloadMsg, "Reload");
+					if (selected === "Reload") {
+						vscode.commands.executeCommand("workbench.action.reloadWindow");
+					}
+				}
+			}
+		),
+		vscode.workspace.onDidChangeWorkspaceFolders(
+			async (event: vscode.WorkspaceFoldersChangeEvent) => {
+				if (event.removed.length > 0) {
+					await stopClients(sortedWorkspaceFolders(event.removed));
+				}
+				if (event.added.length > 0) {
+					await startClients(sortedWorkspaceFolders(event.added));
 				}
 			}
 		)
 	);
 
+	const useLs: boolean = config('terraform').get("languageServer.external");
 	if (useLs) {
-		return installThenStart(context, config);
+		return vscode.commands.executeCommand("terraform.enableLanguageServer");
+	} else {
+		console.log('terraform.languageServer.external is set to false');
 	}
 }
 
-export function deactivate(): Thenable<void> | undefined {
-	if (!client) {
-		return undefined;
-	}
-	return client.stop();
+export function deactivate() {
+	return stopClients();
 }
 
-async function installThenStart(context: vscode.ExtensionContext, config: vscode.WorkspaceConfiguration) {
-	const command: string = config.get("languageServer.pathToBinary");
-	if (command) { // Skip install/upgrade if user has set custom binary path
-		await startLsClient(command, config);
-	} else {
-		const installer = new LanguageServerInstaller;
-		const installDir = `${context.extensionPath}/lsp`;
-
-		try {
-			await installer.install(installDir);
-			await startLsClient(`${installDir}/terraform-ls`, config);
-		} catch (err) {
-			vscode.window.showErrorMessage(err);
-			throw err;
+async function startClients(folders = sortedWorkspaceFolders()) {
+	console.log("Starting:", folders);
+	const command = await pathToBinary();
+	let disposables: vscode.Disposable[] = [];
+	for (const folder of folders) {
+		if (!clients.has(folder)) {
+			const client = newClient(command, folder);
+			disposables.push(client.start());
+			clients.set(folder, client);
+		} else {
+			console.log(`Client for folder: ${folder} already started`);
 		}
 	}
+	return disposables
 }
 
-async function startLsClient(cmd: string, config: vscode.WorkspaceConfiguration) {
+function newClient(cmd: string, folder: string) {
 	const binaryName = cmd.split("/").pop();
-	const lsConfig = vscode.workspace.getConfiguration("terraform-ls");
-	const serverArgs: string[] = config.get("languageServer.args");
-	let serverOptions: ServerOptions;
-	let initializationOptions = { rootModulePaths: lsConfig.get("rootModules") };
+	const channelName = `${binaryName}/${folder}`;
+	const serverArgs: string[] = config('terraform').get('languageServer.args');
+	let initializationOptions = { rootModulePaths: config('terraform-ls').get("rootModules") };
```

Comment:
I feel like with the client side now handling folder context (via this PR)... this argument shouldn't be passed along? Locally testing this config is unset for me, so it hasn't caused any bugs.

Reply:
Are you asking if we want to remove the `rootModulePaths` config entirely?

Reply:
I am unclear on the cases it covers (have not used them myself yet), but maybe?

Reply:
I spoke with Paul and understand this now, I load the config with a folder based scope